src/Microsoft.ML.TensorFlow/TensorFlow/Tensor.cs

```diff
@@ -446,6 +446,104 @@ internal static unsafe TFTensor CreateString(byte[] buffer)
+            // [offset1, offset2,...,offsetn, s1size, s1bytes, s2size, s2bytes,...,snsize,snbytes]
+            //
+            int size = 0;
+            foreach(var b in buffer)
+            {
+                size += (int)TF_StringEncodedSize((UIntPtr)b.Length);
+            }
+            int totalSize = size + buffer.Length * 8;
+            ulong offset = 0;
+            IntPtr handle = TF_AllocateTensor(TFDataType.String, shape.dims, shape.dims.Length, (UIntPtr)totalSize);
+
+            // Clear offset table
+            IntPtr pOffset = TF_TensorData(handle);
+            IntPtr dst = pOffset + buffer.Length * 8;
+            IntPtr dstLimit = pOffset + totalSize;
+            for (int i = 0; i < buffer.Length; i++)
+            {
+                Marshal.WriteInt64(pOffset, (long)offset);
+                using (var status = new TFStatus())
+                {
+                    fixed (byte* src = &buffer[i][0])
+                    {
+                        var written = TF_StringEncode(src, (UIntPtr)buffer[i].Length, (sbyte*)dst, (size_t)(dstLimit.ToInt64() - dst.ToInt64()), status.handle);
+                        var ok = status.StatusCode == TFCode.Ok;
+                        if (!ok)
+                            return null;
+                        pOffset += 8;
+                        dst = (IntPtr)(dst.ToInt64() + (Int64)written);
+                        offset += written.ToUInt64();
+                    }
+                }
+            }
+            return new TFTensor(handle);
+        }
+
+        internal static unsafe byte[][] DecodeStringTensor(TFTensor tensor)
+        {
+            if (tensor == null)
+                throw new ArgumentNullException(nameof(tensor));
+            //
+            // TF_STRING tensors are encoded with a table of 8-byte offsets followed by
+            // [offset1, offset2,...,offsetn, s1size, s1bytes, s2size, s2bytes,...,snsize,snbytes]
+            //
+            long size = 1;
+            foreach (var s in tensor.Shape)
+                size *= s;
+
+            var buffer = new byte[size][];
+            var src = TF_TensorData(tensor.handle);
+            var srcLen = (IntPtr)(src.ToInt64() + (long)tensor.TensorByteSize);
+            src += (int)(size * 8);
+            for (int i = 0; i < buffer.Length; i++)
+            {
+                using (var status = new TFStatus())
+                {
+                    IntPtr dst = IntPtr.Zero;
+                    UIntPtr dstLen = UIntPtr.Zero;
+                    var read = TF_StringDecode((sbyte*)src, (size_t)(srcLen.ToInt64() - src.ToInt64()), (sbyte**)&dst, &dstLen, status.handle);
+                    var ok = status.StatusCode == TFCode.Ok;
+                    if (!ok)
+                        return null;
+                    buffer[i] = new byte[(int)dstLen];
+                    Marshal.Copy(dst, buffer[i], 0, buffer[i].Length);
+                    src = (IntPtr)(src.ToInt64() + (long)read);
```

Comment:
Are we OK with assuming a 64-bit process here?

Reply:
Though its fine to have that assumption because TF is 64bit but I want to it be generic. Is there a way to make it generic?

---
In reply to: [256870163](https://github.com/dotnet/machinelearning/pull/2546#discussion_r256870163) [](ancestors = 256870163)

Reply:
Can you use `IntPtr.Add`?

[https://docs.microsoft.com/en-us/dotnet/api/system.intptr.add?view=netframework-4.7.2](https://docs.microsoft.com/en-us/dotnet/api/system.intptr.add?view=netframework-4.7.2)

---
In reply to: [256960400](https://github.com/dotnet/machinelearning/pull/2546#discussion_r256960400) [](ancestors = 256960400,256870163)

Reply:
or even the `+` operator:

`src += read;`

---
In reply to: [257049147](https://github.com/dotnet/machinelearning/pull/2546#discussion_r257049147) [](ancestors = 257049147,256960400,256870163)